proof/courier.go

```diff
@@ -737,6 +809,77 @@ type HashMailCourier struct {
 	subscriberMtx sync.Mutex
 }
 
+// NewHashMailCourier creates a new hashmail proof courier service handle.
+func NewHashMailCourier(ctx context.Context, cfg *HashMailCourierCfg,
+	transferLog TransferLog, courierAddr *url.URL,
+	lazyConnect bool) (*HashMailCourier, error) {
+
+	courier := HashMailCourier{
+		cfg:           cfg,
+		addr:          courierAddr,
+		transferLog:   transferLog,
+		backoffHandle: NewBackoffHandler(cfg.BackoffCfg, transferLog),
+		subscribers:   make(map[uint64]*fn.EventReceiver[fn.Event]),
+	}
+
+	// If we're not lazy connecting, then we'll attempt to connect to the
+	// mailbox service immediately.
+	if !lazyConnect {
+		err := courier.ensureConnect(ctx)
+		if err != nil {
+			return nil, fmt.Errorf("unable to connect to mailbox "+
+				"service during hashmail courier "+
+				"instantiation: %w", err)
+		}
+	}
+
+	return &courier, nil
+}
+
+// ensureConnect ensures that the courier is connected to the hashmail server.
+// This method does nothing if a mailbox service connection is already
+// established.
+func (h *HashMailCourier) ensureConnect(ctx context.Context) error {
+	// If we're already connected, we'll return early.
+	if h.mailbox != nil {
+		// If the mailbox is already instantiated, we'll check the
+		// to determine if the connection is ready.
+		connStatus, err := h.mailbox.ConnectionStatus()
+		if err != nil {
+			return fmt.Errorf("unable to determine connection "+
+				"status: %w", err)
+		}
+
+		// Return early and don't attempt to establish a new connection
+		// if the connection status is idle, ready, or connecting.
+		if connStatus == CourierConnStatusIdle ||
```

Comment:
**If** we add a new type (see comment above), we might as well add a method `IsPending()` or something like that.